### PR TITLE
Resolve missing internal state update in EventSourcedRememberEntities…

### PR DESCRIPTION
**Pull request by jneumaier**
See #30215

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @jneumaier,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[https://www.lightbend.com/contribute/cla](https://www.lightbend.com/contribute/cla)

**Comment by johanandren** [comment]
OK TO TEST

File: akka-cluster-sharding/src/test/scala/akka/cluster/sharding/internal/RememberEntitiesShardStoreSpec.scala [hunk 1]
```diff
@@ -17,17 +19,21 @@ import org.scalatest.wordspec.AnyWordSpecLike
...
 object RememberEntitiesShardStoreSpec {
-  def config = ConfigFactory.parseString("""
+  def config = ConfigFactory.parseString(s"""
       akka.loglevel=DEBUG
       akka.loggers = ["akka.testkit.SilenceAllTestEventListener"]
       akka.actor.provider = cluster
       akka.remote.artery.canonical.port = 0
       akka.remote.classic.netty.tcp.port = 0
       akka.cluster.sharding.state-store-mode = ddata
+      akka.cluster.sharding.snapshot-after = 2
```
Comment (johanandren) [inline akka-cluster-sharding/src/test/scala/akka/cluster/sharding/internal/RememberEntitiesShardStoreSpec.scala:29]: I think this, and config for a snapshot store below is the only thing that needs to be changed in the test to exercise the bug and that you can skip the added test case.

**Review by johanandren — COMMENTED** [review]
Good catch, curious that nobody noticed earlier.

File: akka-cluster-sharding/src/test/scala/akka/cluster/sharding/internal/RememberEntitiesShardStoreSpec.scala (hunk 1 above)
Comment (jneumaier) [reply akka-cluster-sharding/src/test/scala/akka/cluster/sharding/internal/RememberEntitiesShardStoreSpec.scala:29]: Thanks. I thought the additional test might be easier to read but it does not seem to work with DDataRememberEntitiesShardStore. I removed the test and an unrelated config change.

**Comment by jneumaier** [comment]
@chbatey You also contributed to EventSourcedRememberEntitiesShardStore, maybe you can do a review?

**Review by octonato — APPROVED** [review]
LGTM

**Review by johanandren — APPROVED** [review]

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntitiesShardStore.scala [hunk 2]
```diff
@@ -101,7 +101,7 @@ private[akka] final class EventSourcedRememberEntitiesShardStore(
           left -= 1
           if (left == 0) {
             sender() ! RememberEntitiesShardStore.UpdateDone(started, stopped)
-            state.copy(state.entities.union(started).diff(stopped))
+            state = state.copy(state.entities.union(started).diff(stopped))
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntitiesShardStore.scala:104]: 😱

**Review by patriknw — COMMENTED** [review]
thanks for fixing

**Pull request merged**